src/AutoRest.CSharp/Common/Output/Models/Types/SignatureType.cs

```diff
@@ -0,0 +1,254 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis;
+using AutoRest.CSharp.Input.Source;
+using AutoRest.CSharp.Mgmt.AutoRest;
+using AutoRest.CSharp.Output.Models.Shared;
+using System;
+using System.Linq;
+using AutoRest.CSharp.Mgmt.Decorator;
+using System.Diagnostics.CodeAnalysis;
+
+namespace AutoRest.CSharp.Output.Models.Types
+{
+    /// <summary>
+    /// This type holds three portions of codes:
+    ///     - current
+    ///     - custom
+    ///     - baseline contract
+    ///     current union custom compare with baseline contract outputs the changeset, we can apply different rules with it.
+    /// </summary>
+    internal class SignatureType
+    {
+        private readonly string _defaultNamespace;
+        private readonly string _defaultName;
+        private readonly SignatureType? _customization;
+        private readonly SignatureType? _baselineContract;
```

Comment:
I am confused why we have to make this type recursive?
For instance, do we ever need the `OverloadMethods` property on this `_customization` field?
It feels like we do not need everything on SignatureType for customization - then maybe we should not choose `SignatureType` as the type of `customization`.

Reply:
For example, I only see `Methods` property is used on `_customization`.
Does it mean this `_customization` should just be `_customizationMethods` with a type of `IReadOnlyList<MethodSignature>`?
Same comment applies to `_baselineContract` as well

Reply:
Yes, only `Methods` is in use for now. We will have `Properties` later, so there will be `_customMethods` and `_customProperties`, it grows every time we add a new member.
It would be more convenient to hold them in the same type for custom recursively.